### PR TITLE
Move semantics solutions

### PR DESCRIPTION
**Pull request by TonySchneider**

### CODE REVIEW

File: exercises/if/if2.rs [hunk 1]
```diff
@@ -5,7 +5,7 @@
 //
 // Execute `rustlings hint if2` or use the `hint` watch subcommand for a hint.
 
-// I AM NOT DONE
+
```
Comment (MatanHamilis) [inline exercises/if/if2.rs:8]: Notice when committing these sneaky newlines.

File: exercises/move_semantics/move_semantics1.rs [hunk 2]
```diff
@@ -3,19 +3,18 @@
...
 fn main() {
     let vec0 = vec![22, 44, 66];
 
     let vec1 = fill_vec(vec0);
-
+    
     assert_eq!(vec1, vec![22, 44, 66, 88]);
 }
 
-fn fill_vec(mut vec: Vec<i32>) -> Vec<i32> {
```
Comment (MatanHamilis) [inline exercises/move_semantics/move_semantics1.rs:17]: Hmm... whats wrong with the original solution you deleted? Why did you define another variable?

File: exercises/move_semantics/move_semantics2.rs [hunk 3]
```diff
@@ -5,15 +5,17 @@
 // Execute `rustlings hint move_semantics2` or use the `hint` watch subcommand
 // for a hint.
 
-// I AM NOT DONE
-
 #[test]
 fn main() {
     let vec0 = vec![22, 44, 66];
+    assert_eq!(vec0, vec![22, 44, 66]);
```
Comment (MatanHamilis) [inline exercises/move_semantics/move_semantics2.rs]: Why did you add this line?

File: exercises/move_semantics/move_semantics2.rs [hunk 4]
```diff
@@ -5,15 +5,17 @@
...
 
     let mut vec1 = fill_vec(vec0);
 
-    assert_eq!(vec0, vec![22, 44, 66]);
+    // the ownership of vec0 has passed to the argument of the function so it was removed so we can print vec0 here.
```
Comment (MatanHamilis) [inline exercises/move_semantics/move_semantics2.rs]: Can you think of a way such that `vec0` won't be moved? (hint: a reference doesn't move a variable...)

File: exercises/move_semantics/move_semantics3.rs [hunk 5]
```diff
@@ -17,7 +15,8 @@ fn main() {
     assert_eq!(vec1, vec![22, 44, 66, 88]);
 }
 
-fn fill_vec(vec: Vec<i32>) -> Vec<i32> {
+// I have added "mut" since we are changing the value of vec inside the function.
```
Comment (MatanHamilis) [inline exercises/move_semantics/move_semantics3.rs:18]: Great!

File: exercises/move_semantics/move_semantics5.rs [hunk 6]
```diff
@@ -6,14 +6,14 @@
...
-// I AM NOT DONE
-
 #[test]
 fn main() {
     let mut x = 100;
     let y = &mut x;
-    let z = &mut x;
     *y += 100;
+    // We can't define two mutable variables by using references because we want to avoid changing the value of X simultaneously.
+    // Therefore, we must use 'y' before we borrow 'x' a second time.
```
Comment (MatanHamilis) [inline exercises/move_semantics/move_semantics5.rs:15]: Great!

File: exercises/move_semantics/move_semantics6.rs [hunk 7]
```diff
@@ -5,24 +5,22 @@
 // Execute `rustlings hint move_semantics6` or use the `hint` watch subcommand
 // for a hint.
 
-// I AM NOT DONE
-
 fn main() {
     let data = "Rust is great!".to_string();
 
-    get_char(data);
+    let last_char = get_char(&data);
```
Comment (MatanHamilis) [inline exercises/move_semantics/move_semantics6.rs]: why did you add a variable definition here?

File: exercises/quiz1.rs [hunk 8]
```diff
@@ -13,7 +13,7 @@
 //
 // No hints this time ;)
 
-// I AM NOT DONE
```
Comment (MatanHamilis) [inline exercises/quiz1.rs:16]: The code in this quiz is incorrect!

- [ ] Fix the code

File: exercises/vecs/vecs1.rs [hunk 9]
```diff
@@ -18,7 +17,7 @@ fn array_and_vec() -> ([i32; 4], Vec<i32>) {
     // v.push(30);
     // v.push(40);
 
-    // let v: Vec<i32> = Vec::from(a);
+    let v: Vec<i32> = Vec::from(a);
```
Comment (MatanHamilis) [inline exercises/vecs/vecs1.rs]: Heard of [`slice::to_vec()`](https://doc.rust-lang.org/std/primitive.slice.html#method.to_vec)?

Comment (MatanHamilis) [reply exercises/vecs/vecs1.rs]: Also: Does "from" take a `&Vec` (ref) or does it take ownership of a vector? You pass it `a` but you use `a` later. Can you explain this?

File: exercises/vecs/vecs2.rs [hunk 10]
```diff
@@ -7,7 +7,6 @@
 //
 // Execute `rustlings hint vecs2` or use the `hint` watch subcommand for a hint.
 
-// I AM NOT DONE
```
Comment (MatanHamilis) [inline exercises/vecs/vecs2.rs:10]: - [ ] Use [`Iterator::for_each()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each)

**Review by MatanHamilis — CHANGES_REQUESTED** [review]
Great work!
Not a lot of fixes, let me know when you're ready for another review.

File: exercises/move_semantics/move_semantics1.rs (hunk 2 above)
Comment (TonySchneider) [reply exercises/move_semantics/move_semantics1.rs:17]: actually, I can leave this "mut" in the argument since the ownership is moved to the variable inside the function and only after that am I changing the value so it doesn't matter whether mut is there or not

File: exercises/if/if2.rs (hunk 1 above)
Comment (TonySchneider) [reply exercises/if/if2.rs:8]: Ok, will notice in the future

File: exercises/move_semantics/move_semantics2.rs (hunk 4 above)
Comment (TonySchneider) [reply exercises/move_semantics/move_semantics2.rs]: yep, just send to the function a reference of vec0 and vec0 will stay with its value since the ownership is stay there and doesn't move to the function

Comment (TonySchneider) [reply exercises/move_semantics/move_semantics2.rs]: but I think I will need to do some "clone()" inside the function when I am defining there a new variable.

Comment (TonySchneider) [reply exercises/move_semantics/move_semantics2.rs]:
```
fn main() {
    let vec0 = vec![22, 44, 66];
    // assert_eq!(vec0, vec![22, 44, 66]);

    let mut vec1 = fill_vec(&vec0);

    assert_eq!(vec0, vec![22, 44, 66]);
    assert_eq!(vec1, vec![22, 44, 66, 88]);
}

fn fill_vec(vec: &Vec<i32>) -> Vec<i32> {
    let mut vec = vec.clone();

    vec.push(88);

    vec
}
```

File: exercises/move_semantics/move_semantics6.rs (hunk 7 above)
Comment (TonySchneider) [reply exercises/move_semantics/move_semantics6.rs]: I added it automatically since I thought I would use it in the next steps but found that I didn't need it.  I will remove it

File: exercises/move_semantics/move_semantics1.rs (hunk 2 above)
Comment (MatanHamilis) [reply exercises/move_semantics/move_semantics1.rs:17]: But why do we need a variable inside the function?

Comment (TonySchneider) [reply exercises/move_semantics/move_semantics1.rs:17]: It was there, I didn't add it

File: exercises/quiz1.rs (hunk 8 above)
Comment (TonySchneider) [reply exercises/quiz1.rs:16]: that's what happens when you rush and don't read the instructions at the beginning of the file.
fixed it..

File: exercises/vecs/vecs1.rs (hunk 9 above)
Comment (TonySchneider) [reply exercises/vecs/vecs1.rs]: Yes, when I am passing this array to Vec::from() it is cloning this array to a new array and converting it to vector (something like this).
For example:
If I will do this:
let c = a;
so it will copy the array into c and not move ownership. because it is a type that has the copy trait like integers, booleans, etc..

Comment (TonySchneider) [reply exercises/vecs/vecs1.rs]: regarding slice::to_vec(), I can convert a part of an array to a Vec or even the entire array. but is a.to_vec() better in this situation than using Vec::from(a)? @MatanHamilis

File: exercises/vecs/vecs2.rs (hunk 10 above)
Comment (TonySchneider) [reply exercises/vecs/vecs2.rs:10]: changed it

File: exercises/vecs/vecs1.rs (hunk 9 above)
Comment (MatanHamilis) [reply exercises/vecs/vecs1.rs]: I find it more readable, but just so you know both options exist.

File: exercises/move_semantics/move_semantics2.rs (hunk 3 above)
Comment (TonySchneider) [reply exercises/move_semantics/move_semantics2.rs]: I changed it, you need to look at the latest files here: https://github.com/TonySchneider/rustlings-solutions/pull/1/files
